### ISSUE TITLE
Unable to insert VARBYTE via JDBC prepared statements

### ISSUE THREAD

**Issue by hashhar** [aws/amazon-redshift-jdbc-driver#39]
## Driver version
2.1.0.1

## Redshift version
PostgreSQL 8.0.2 on i686-pc-linux-gnu, compiled by GCC gcc (GCC) 3.4.2 20041017 (Red Hat 3.4.2-6.fc3), Redshift 1.0.33595��

## Client Operating System
MacOS Monterey

## JAVA/JVM version
openjdk version "11.0.13" 2021-10-19 LTS
OpenJDK Runtime Environment Zulu11.52+13-CA (build 11.0.13+8-LTS)
OpenJDK 64-Bit Server VM Zulu11.52+13-CA (build 11.0.13+8-LTS, mixed mode)

## Table schema
```sql
CREATE TABLE test (vb varbyte);
```

## Problem description

1. Expected behaviour: I should be able to insert values to VARBYTE columns
2. Actual behaviour: Unable to insert values to VARBYTE columns
3. Error message/stack trace: 
  ```
  com.amazon.redshift.util.RedshiftException: ERROR: column "vb" is of type binary varying but expression is of type bytea
    Hint: You will need to rewrite or cast the expression.
	  at com.amazon.redshift.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2601)
	  at com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread(QueryExecutorImpl.java:2269)
	  at com.amazon.redshift.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1880)
	  at com.amazon.redshift.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1872)
	  at com.amazon.redshift.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:368)
	  at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeInternal(RedshiftStatementImpl.java:514)
	  at com.amazon.redshift.jdbc.RedshiftStatementImpl.execute(RedshiftStatementImpl.java:435)
	  at com.amazon.redshift.jdbc.RedshiftPreparedStatement.executeWithFlags(RedshiftPreparedStatement.java:200)
	  at com.amazon.redshift.jdbc.RedshiftPreparedStatement.execute(RedshiftPreparedStatement.java:184)
	  at Scratch.main(scratch.java:23)
  ```
4. Any other details that can be helpful:

## JDBC trace logs
```
Nov 23 18:31:22.778 DEBUG  [1 main] com.amazon.redshift.Driver.connect: ===================================
Nov 23 18:31:22.778 FUNCTION  [1 main] com.amazon.redshift.Driver.connect:  Enter (jdbc:redshift://redshift-testing-cluster.redacted;DSILogLevel=6;LogPath=/tmp/redshift.log,{password=***, user=redacted}) 
Nov 23 18:31:22.778 DEBUG  [1 main] com.amazon.redshift.Driver.connect: Connecting with URL: jdbc:redshift://redshift-testing-cluster.redacted;DSILogLevel=6;LogPath=/tmp/redshift.log
Nov 23 18:31:22.788 DEBUG  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.RedshiftConnectionImpl: Redshift JDBC Driver 2.1.0.1
Nov 23 18:31:22.790 DEBUG  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.setDefaultFetchSize:   setDefaultFetchSize = 0
Nov 23 18:31:22.790 DEBUG  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.setPrepareThreshold:   setPrepareThreshold = 5
Nov 23 18:31:22.800 DEBUG  [1 main] com.amazon.redshift.core.v3.ConnectionFactoryImpl.openConnectionImpl: Trying to establish a protocol version 3 connection to redshift-testing-cluster.redacted:10642
Nov 23 18:31:23.380 INFO  [1 main] com.amazon.redshift.core.RedshiftStream.RedshiftStream: Gets a new stream on a new socket
Nov 23 18:31:23.382 DEBUG  [1 main] com.amazon.redshift.core.v3.ConnectionFactoryImpl.tryConnect: Receive Buffer Size is 132,090
Nov 23 18:31:23.383 DEBUG  [1 main] com.amazon.redshift.core.v3.ConnectionFactoryImpl.tryConnect: Send Buffer Size is 132,090
Nov 23 18:31:23.383 DEBUG  [1 main] com.amazon.redshift.core.v3.ConnectionFactoryImpl.enableSSL:  FE=> SSLRequest
Nov 23 18:31:23.672 DEBUG  [1 main] com.amazon.redshift.core.v3.ConnectionFactoryImpl.enableSSL:  <=BE SSLOk
Nov 23 18:31:24.353 DEBUG  [1 main] com.amazon.redshift.core.v3.ConnectionFactoryImpl.sendStartupPacket:  FE=> StartupPacket(user=redacted, database=testdb, client_encoding=UTF8, DateStyle=ISO, TimeZone=Asia/Kolkata, extra_float_digits=2, driver_version=Redshift JDBC Driver 2.1.0.1, os_version=Mac OS X 12.0.1 x86_64, client_protocol_version=3)
Nov 23 18:31:24.628 DEBUG  [1 main] com.amazon.redshift.core.v3.ConnectionFactoryImpl.doAuthentication:  <=BE AuthenticationReqMD5
Nov 23 18:31:24.630 DEBUG  [1 main] com.amazon.redshift.core.v3.ConnectionFactoryImpl.doAuthentication:  FE=> Password(md5digest)
Nov 23 18:31:24.908 DEBUG  [1 main] com.amazon.redshift.core.v3.ConnectionFactoryImpl.doAuthentication:  <=BE AuthenticationOk
Nov 23 18:31:24.935 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(application_name = unset)
Nov 23 18:31:24.935 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(case_sensitive = on)
Nov 23 18:31:24.936 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(client_encoding = UTF8)
Nov 23 18:31:24.936 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(datestyle = ISO, MDY)
Nov 23 18:31:24.936 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(integer_datetimes = on)
Nov 23 18:31:24.936 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(is_superuser = on)
Nov 23 18:31:24.937 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(server_encoding = UNICODE)
Nov 23 18:31:24.937 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(server_version = 8.0.2)
Nov 23 18:31:24.937 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(session_authorization = redacted)
Nov 23 18:31:24.937 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(TimeZone = Asia/Kolkata)
Nov 23 18:31:24.942 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(timezone_abbreviations = Default)
Nov 23 18:31:24.942 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(padb_version = 1.0.33595)
Nov 23 18:31:24.943 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(padb_revision = cd15d6dca8)
Nov 23 18:31:24.943 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(max_numeric_precision = 38)
Nov 23 18:31:24.943 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(max_varchar_size = 65535)
Nov 23 18:31:24.943 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(datashare_enabled = off)
Nov 23 18:31:24.943 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(session_id = 5d4f0da4-2f28-461c-99e6-81c740551135)
Nov 23 18:31:24.943 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveParameterStatus:  <=BE ParameterStatus(server_protocol_version = 2)
Nov 23 18:31:24.943 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.readStartupMessages:  <=BE BackendKeyData(pid=1,073,881,358,ckey=74,892,290)
Nov 23 18:31:24.944 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveRFQ:  <=BE ReadyForQuery(I)
Nov 23 18:31:24.946 DEBUG  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.RedshiftConnectionImpl:     types using binary send = TIMESTAMPTZ,UUID,INT2_ARRAY,INT4_ARRAY,BYTEA,TEXT_ARRAY,TIMETZ,INT8,INT2,INT4,VARCHAR_ARRAY,INT8_ARRAY,POINT,TIMESTAMP,TIME,BOX,FLOAT4,FLOAT8,FLOAT4_ARRAY,FLOAT8_ARRAY
Nov 23 18:31:24.946 DEBUG  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.RedshiftConnectionImpl:     types using binary receive = TIMESTAMPTZ,UUID,INT2_ARRAY,INT4_ARRAY,BYTEA,TEXT_ARRAY,TIMETZ,INT8,INT2,INT4,VARCHAR_ARRAY,INT8_ARRAY,POINT,DATE,TIMESTAMP,TIME,BOX,FLOAT4,FLOAT8,FLOAT4_ARRAY,FLOAT8_ARRAY
Nov 23 18:31:24.947 DEBUG  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.RedshiftConnectionImpl:     integer date/time = true
Nov 23 18:31:24.951 FUNCTION  [1 main] com.amazon.redshift.Driver.connect:  Return com.amazon.redshift.jdbc.RedshiftConnectionImpl@134d26af 
Nov 23 18:31:24.952 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Enter () 
Nov 23 18:31:24.952 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Enter (1003,1007) 
Nov 23 18:31:24.952 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Enter (1003,1007,2) 
Nov 23 18:31:24.956 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.setFetchSize:  Enter (0) 
Nov 23 18:31:24.956 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.setFetchSize:  Return  
Nov 23 18:31:24.956 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Return com.amazon.redshift.jdbc.RedshiftStatementImpl@5adb0db3 
Nov 23 18:31:24.956 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Return com.amazon.redshift.jdbc.RedshiftStatementImpl@5adb0db3 
Nov 23 18:31:24.957 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Return com.amazon.redshift.jdbc.RedshiftStatementImpl@5adb0db3 
Nov 23 18:31:24.957 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.execute:  Enter (DROP TABLE IF EXISTS ashhar_test) 
Nov 23 18:31:24.957 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.getAutoCommit:  Enter () 
Nov 23 18:31:24.957 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.getAutoCommit:  Return true 
Nov 23 18:31:24.958 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.execute:   simple execute, handler=com.amazon.redshift.jdbc.RedshiftStatementImpl$StatementResultHandler@253d9f73, maxRows=0, fetchSize=0, flags=17
Nov 23 18:31:24.959 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendParse:  FE=> Parse(stmt=null,query="DROP TABLE IF EXISTS ashhar_test",oids={})
Nov 23 18:31:24.959 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendBind:  FE=> Bind(stmt=null,portal=null)
Nov 23 18:31:24.959 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendDescribePortal:  FE=> Describe(portal=null)
Nov 23 18:31:24.959 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendExecute:  FE=> Execute(portal=null,limit=0)
Nov 23 18:31:24.960 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendFlush:  FE=> Flush
Nov 23 18:31:24.960 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendSync:  FE=> Sync
Nov 23 18:31:25.261 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread:  <=BE ParseComplete [null]
Nov 23 18:31:25.261 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread:  <=BE BindComplete [unnamed]
Nov 23 18:31:25.261 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread:  <=BE NoData
Nov 23 18:31:25.262 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveCommandStatus:  <=BE CommandStatus(DROP TABLE)
Nov 23 18:31:25.263 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveRFQ:  <=BE ReadyForQuery(I)
Nov 23 18:31:25.264 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.execute:  Return  
Nov 23 18:31:25.264 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Enter () 
Nov 23 18:31:25.264 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Enter (1003,1007) 
Nov 23 18:31:25.264 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Enter (1003,1007,2) 
Nov 23 18:31:25.264 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.setFetchSize:  Enter (0) 
Nov 23 18:31:25.265 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.setFetchSize:  Return  
Nov 23 18:31:25.265 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Return com.amazon.redshift.jdbc.RedshiftStatementImpl@324a0017 
Nov 23 18:31:25.265 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Return com.amazon.redshift.jdbc.RedshiftStatementImpl@324a0017 
Nov 23 18:31:25.265 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.createStatement:  Return com.amazon.redshift.jdbc.RedshiftStatementImpl@324a0017 
Nov 23 18:31:25.265 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.execute:  Enter (CREATE TABLE ashhar_test (vb varbyte)) 
Nov 23 18:31:25.265 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.getAutoCommit:  Enter () 
Nov 23 18:31:25.266 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.getAutoCommit:  Return true 
Nov 23 18:31:25.266 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.execute:   simple execute, handler=com.amazon.redshift.jdbc.RedshiftStatementImpl$StatementResultHandler@6187d1f5, maxRows=0, fetchSize=0, flags=17
Nov 23 18:31:25.266 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendParse:  FE=> Parse(stmt=null,query="CREATE TABLE ashhar_test (vb varbyte)",oids={})
Nov 23 18:31:25.266 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendBind:  FE=> Bind(stmt=null,portal=null)
Nov 23 18:31:25.266 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendDescribePortal:  FE=> Describe(portal=null)
Nov 23 18:31:25.267 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendExecute:  FE=> Execute(portal=null,limit=0)
Nov 23 18:31:25.267 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendFlush:  FE=> Flush
Nov 23 18:31:25.267 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendSync:  FE=> Sync
Nov 23 18:31:25.568 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread:  <=BE ParseComplete [null]
Nov 23 18:31:25.568 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread:  <=BE BindComplete [unnamed]
Nov 23 18:31:25.568 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread:  <=BE NoData
Nov 23 18:31:25.569 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveCommandStatus:  <=BE CommandStatus(CREATE TABLE)
Nov 23 18:31:25.574 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveRFQ:  <=BE ReadyForQuery(I)
Nov 23 18:31:25.574 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.execute:  Return  
Nov 23 18:31:25.574 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.prepareStatement:  Enter (INSERT INTO ashhar_test (vb) VALUES (?)) 
Nov 23 18:31:25.575 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.prepareStatement:  Enter (INSERT INTO ashhar_test (vb) VALUES (?),1003,1007) 
Nov 23 18:31:25.575 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.prepareStatement:  Enter (INSERT INTO ashhar_test (vb) VALUES (?),1003,1007,2) 
Nov 23 18:31:25.579 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.setFetchSize:  Enter (0) 
Nov 23 18:31:25.579 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.setFetchSize:  Return  
Nov 23 18:31:25.579 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.setPoolable:  Enter (true) 
Nov 23 18:31:25.580 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.prepareStatement:  Return INSERT INTO ashhar_test (vb) VALUES (?) 
Nov 23 18:31:25.580 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.prepareStatement:  Return INSERT INTO ashhar_test (vb) VALUES (?) 
Nov 23 18:31:25.580 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.prepareStatement:  Return INSERT INTO ashhar_test (vb) VALUES (?) 
Nov 23 18:31:25.580 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftPreparedStatement.setBytes:  Enter (1,[B@63798ca7) 
Nov 23 18:31:25.581 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftPreparedStatement.execute:  Enter () 
Nov 23 18:31:25.582 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.getAutoCommit:  Enter () 
Nov 23 18:31:25.582 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.getAutoCommit:  Return true 
Nov 23 18:31:25.582 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.execute:   simple execute, handler=com.amazon.redshift.jdbc.RedshiftStatementImpl$StatementResultHandler@32f232a5, maxRows=0, fetchSize=0, flags=16
Nov 23 18:31:25.582 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendParse:  FE=> Parse(stmt=S_1-20625376990060,query="INSERT INTO ashhar_test (vb) VALUES ($1)",oids={17})
Nov 23 18:31:25.582 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendBind:  FE=> Bind(stmt=S_1-20625376990060,portal=C_2-20625377211861,$1=<?>,type=BYTEA)
Nov 23 18:31:25.583 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendDescribePortal:  FE=> Describe(portal=C_2-20625377211861)
Nov 23 18:31:25.583 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendExecute:  FE=> Execute(portal=C_2-20625377211861,limit=0)
Nov 23 18:31:25.583 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendFlush:  FE=> Flush
Nov 23 18:31:25.583 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendSync:  FE=> Sync
Nov 23 18:31:25.982 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveErrorResponse:  <=BE ErrorMessage(ERROR: column "vb" is of type binary varying but expression is of type bytea
  Hint: You will need to rewrite or cast the expression.
  Location: File: /home/ec2-user/padb/src/pg/src/backend/parser/parse_target.c, Routine: updateTargetListEntry, Line: 658
  Server SQLState: 42804)
Nov 23 18:31:25.984 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveRFQ:  <=BE ReadyForQuery(I)
Nov 23 18:31:25.984 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread:  FE marking setPortalDescribed(false) for query INSERT INTO ashhar_test (vb) VALUES (?)
Nov 23 18:31:25.984 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread:  FE marking setPortalDescribed(false) for query SYNC
Nov 23 18:31:25.985 ERROR  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.rollbackIfRequired: com.amazon.redshift.util.RedshiftException: ERROR: column "vb" is of type binary varying but expression is of type bytea
  Hint: You will need to rewrite or cast the expression.
	at com.amazon.redshift.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2601)
	at com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread(QueryExecutorImpl.java:2269)
	at com.amazon.redshift.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1880)
	at com.amazon.redshift.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1872)
	at com.amazon.redshift.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:368)
	at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeInternal(RedshiftStatementImpl.java:514)
	at com.amazon.redshift.jdbc.RedshiftStatementImpl.execute(RedshiftStatementImpl.java:435)
	at com.amazon.redshift.jdbc.RedshiftPreparedStatement.executeWithFlags(RedshiftPreparedStatement.java:200)
	at com.amazon.redshift.jdbc.RedshiftPreparedStatement.execute(RedshiftPreparedStatement.java:184)
	at Scratch.main(scratch.java:20)

Nov 23 18:31:25.985 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.close:  Enter () 
Nov 23 18:31:25.985 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.cancel:  Enter () 
Nov 23 18:31:25.985 ERROR  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.cancel: statementState is StatementCancelState.IDLE
Nov 23 18:31:25.985 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.cancel:  Return  
Nov 23 18:31:25.985 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendCloseStatement:  FE=> CloseStatement(S_1-20625376990060)
Nov 23 18:31:25.986 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendFlush:  FE=> Flush
Nov 23 18:31:25.986 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.sendSync:  FE=> Sync
Nov 23 18:31:26.274 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.processSyncOnClose:  <=BE CloseComplete
Nov 23 18:31:26.274 DEBUG  [1 main] com.amazon.redshift.core.v3.QueryExecutorImpl.receiveRFQ:  <=BE ReadyForQuery(I)
Nov 23 18:31:26.274 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftStatementImpl.close:  Return  
Nov 23 18:31:26.274 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.close:  Enter () 
Nov 23 18:31:26.275 DEBUG  [1 main] com.amazon.redshift.core.QueryExecutorBase.close:  FE=> Terminate
Nov 23 18:31:26.275 INFO  [1 main] com.amazon.redshift.core.RedshiftStream.close: Stream on a connected socket closed
Nov 23 18:31:26.278 FUNCTION  [1 main] com.amazon.redshift.jdbc.RedshiftConnectionImpl.close:  Return  
```

## Reproduction code

```java
import java.nio.charset.StandardCharsets;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;

class Scratch
{
    public static void main(String[] args)
    {
        try (Connection conn = DriverManager.getConnection("jdbc:redshift://<cluster>", "<user>", "<password>")) {
            conn.createStatement().execute("DROP TABLE IF EXISTS test");
            conn.createStatement().execute("CREATE TABLE test (vb varbyte)");
            try (PreparedStatement ps = conn.prepareStatement("INSERT INTO test (vb) VALUES (?)")) {
                ps.setBytes(1, "ashhar".getBytes(StandardCharsets.UTF_8));
                //ps.setObject(1, "ashhar".getBytes(StandardCharsets.UTF_8), Types.LONGVARBINARY);
                //ps.setObject(1, "ashhar".getBytes(StandardCharsets.UTF_8), Types.VARBINARY);
                ps.execute();
            }
            try (ResultSet rs = conn.createStatement().executeQuery("SELECT * FROM test")) {
                while (rs.next()) {
                    System.out.println(rs.getString(1));
                }
            }
        }
        catch (SQLException throwables) {
            throwables.printStackTrace();
        }
}
```

None of the above 3 ways of setting the values work. `setBytes`, `setObject` with LONGVARBINARY or VARBINARY.

**Comment by hashhar** [comment]
Note that adding a cast as suggested `CAST(? AS VARBYTE)` still fails with:

```
com.amazon.redshift.util.RedshiftException: ERROR: cannot cast type bytea to binary varying
	at com.amazon.redshift.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2601)
	at com.amazon.redshift.core.v3.QueryExecutorImpl.processResultsOnThread(QueryExecutorImpl.java:2269)
	at com.amazon.redshift.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1880)
	at com.amazon.redshift.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1872)
	at com.amazon.redshift.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:368)
	at com.amazon.redshift.jdbc.RedshiftStatementImpl.executeInternal(RedshiftStatementImpl.java:514)
	at com.amazon.redshift.jdbc.RedshiftStatementImpl.execute(RedshiftStatementImpl.java:435)
	at com.amazon.redshift.jdbc.RedshiftPreparedStatement.executeWithFlags(RedshiftPreparedStatement.java:200)
	at com.amazon.redshift.jdbc.RedshiftPreparedStatement.execute(RedshiftPreparedStatement.java:184)
```

**Comment by iggarish** [comment]
Thanks for reporting issue. We are looking into it.

**Comment by hashhar** [comment]
A workaround for other people running into this is to use the `to_varbyte` function. It'd be nice to see the driver support writing the VARBYTE values directly in the long term though.

e.g. use `to_varbyte(?, 'hex')` in your SQL.

And bind the parameter using something like `ps.setString(1, base16().encode(value.getBytes()))` (base16 is from `com.google.common.io.BaseEncoding`).

Or if you are sure your values are all valid UTF-8 strings then use `to_varbyte(?, 'utf-8')` in SQL and bind parameter directly like `ps.setString(1, validUtf8StringValue)`.

**Comment by iggarish** [comment]
Thanks for the sending the work around. We are discussing the issue to support as BINARY binding.

**Comment by hashhar** [comment]
Thanks for the update, I'll keep an eye out for both the driver and server updates.

If I understand correctly both the driver and server versions need to be updated for the new `setVarbyte` API to work?

**Comment by iggarish** [comment]
>>If I understand correctly both the driver and server versions need to be updated for the new setVarbyte API to work?
Yes that's correct.

**Comment by iggarish** [comment]
Released 2.1.0.4
